File: dan/Problems/Medium/LinkedList/138. Copy List with Random Pointer/solution.py
# Definition for singly-linked list with a random pointer.
# class RandomListNode(object):
#     def __init__(self, x):
#         self.label = x
#         self.next = None
#         self.random = None

class Solution(object):
    #O(n) space complexity
    def copyRandomList1(self, head):
        """
        :type head: RandomListNode
        :rtype: RandomListNode
        """
        dic = {}
        node = head
        while node:
            dic[node] = RandomListNode(node.label)
            node = node.next
        node = head
        while node:
            dic[node].next = dic.get(node.next)
            dic[node].random = dic.get(node.random)
            node = node.next
        return dic.get(head)  
    
    
    #O(1) space cmoplexity
    def copyRandomList(self, head):
        if not head:
            return None
        
        #insert a node after each node in the list
        cur = head
        while cur:
            nxt = cur.next
            cur.next = RandomListNode(cur.label)
            cur.next.next, cur = nxt, nxt
        cur = head
        
        #Allocate random pointers
        while cur:
            if cur.random:
                cur.next.random = cur.random.next
            cur = cur.next.next
            
        #Seperate two lists
        cur = second = head.next
        while cur and cur.next:
            head.next = cur.next
            cur.next = cur.next.next
            cur, head = cur.next, head.next
        head.next = None
            
        return second

File: dan/Problems/Medium/LinkedList/138. Copy List with Random Pointer/solution.java
/**
 * Definition for singly-linked list with a random pointer.
 * class RandomListNode {
 *     int label;
 *     RandomListNode next, random;
 *     RandomListNode(int x) { this.label = x; }
 * };
 */
public class Solution {
    //Use dictionary
    public RandomListNode copyRandomList1(RandomListNode head) {
        if(head == null)
            return null;
        HashMap<RandomListNode, RandomListNode> copy = new HashMap<RandomListNode, RandomListNode>();
        RandomListNode cur = head;
        RandomListNode head2 = new RandomListNode(head.label);
        RandomListNode cur2 = head2;
        while(cur != null){
            copy.put(cur, cur2);
            if (cur.next != null){
                cur2.next = new RandomListNode(cur.next.label);
                cur2 = cur2.next;
            }
            cur = cur.next; 
        }
        cur = head;
        while(cur != null){
            if(cur.random != null)
                copy.get(cur).random = copy.get(cur.random);
            cur = cur.next;
        }
        return head2;
    }
    
    //Add a node after each node
    public RandomListNode copyRandomList(RandomListNode head){
        if(head == null){
            return head;
        }
        RandomListNode cur = head;
        while(cur != null){
            RandomListNode nxt = cur.next;
            cur.next = new RandomListNode(cur.label);
            cur.next.next = nxt;
            cur = nxt;
        }
        cur = head;
        while(cur != null){
            if(cur.random != null)
                cur.next.random = cur.random.next;
            cur = cur.next.next;
        }
        cur = head;
        RandomListNode head2 = head.next;
        RandomListNode cur2 = head2;
        while(cur != null){
            cur.next = cur.next.next;
            cur = cur.next;
            if(cur != null){
                cur2.next = cur2.next.next;
                cur2 = cur2.next;
            }

        }
        return head2;
    }
}
